packages/plugin_shopper_context/cartridges/plugin_shopper_context/cartridge/scripts/helpers/utils.js

```diff
@@ -0,0 +1,63 @@
+'use strict';
+
+var { ALLOW_UNKNOWN_CLIENTS } = require('*/cartridge/scripts/config/constant');
+
+var {
+    getClientRegistry
+} = require('*/cartridge/scripts/config/clientRegistry');
+
+/**
+ * @typedef {Object} ShopperContext
+ * @property {string} sourceCode
+ * @property {string[]} customerGroupIds
+ * @property {string} effectiveDateTime
+ * @property {Object} customQualifiers
+ * @property {Object} assignmentQualifiers
+ */
+
+/**
+ * Validates the given shopper context object against the client registry.
+ * @param {string} clientId - client id to be validated against
+ * @param {ShopperContext} shopperContext - shopper context to be validated
+ * @return {boolean} - validation status
+ */
+function validateContext(clientId, shopperContext) {
+    const clientRegistry = getClientRegistry();
+
+    if (!clientRegistry[clientId]) {
+        // client not in the registry
+        // this means there will be no constraint on shopper context. Use this with caution if you
+        // are using shopperJWT to set context
+        return ALLOW_UNKNOWN_CLIENTS;
+    }
+
+    const clientRules = clientRegistry[clientId];
+    // validate shopperContext against the allowed attributes
+    if (!clientRules.allowSourceCode && shopperContext.sourceCode) {
```

Comment:
Just a note: customers may choose customize the validation logic further. Maybe something like allowing specific known values for `sourceCode`.

So we cannot assume that the customers would configure only the clientRegistry file. I imagine they'd want to customize this validation logic too.

Reply:
Customer will have to get a copy of this plugin and deploy it to BM. Technically they can customise whatever in here, but the minimal they need to do is the client-registry.